### PR TITLE
Unit test of SubscriberListPresenter, SubscriberDetailsPresenter, WebClientPresenter

### PR DESCRIPTION
**Pull request by jvc931**
Unit test of SubscriberListPresenter, SubscriberDetailsPresenter, WebClientPresenter

### CODE REVIEW

File: app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    /**
+     * Unit test of fetchSubscriberDetails makes the call to profileInteractor execute method,
+     * will have response a exception, first should show the loading indicator with the profileInteractor
+     * response should dismiss the loading indicator and show the error message
+     *
+     * @throws Exception
+     */
+    @Test
+    public void fetchSubscriberDetails_withAttachedViewAndProfileInteractorErrorResponse_ShouldShowLoadingIndicatorAndErrorMessage_ShouldNotCallRepositoriesInteractor() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java]: *fetchSubscriberDetails_withViewAttachedAndProfileCallFails_ShouldRenderLoadingIndicatorAndError*

File: app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    /**
+     * Unit test of fetchSubscriberDetails makes the call to profileInteractor execute method,
+     * will have response the profile information with has this response calls the repositoryInteractor
+     * execute method that will have response a exception, first should show the loading
+     * indicator with the repositoryInteractor response dismiss the loading indicator and show the error message
+     *
+     * @throws Exception
+     */
+    @Test
+    public void fetchSubscriberDetails_withAttachedViewAndProfileInteractorResponse_ShouldCallRepositoriesInteractor_withRepositoriesInteractorErrorResponse() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java]: *fetchSubscriberDetails_withViewAttachedAndRepositoriesCallFails_ShouldRenderLoadingIndicatorAndError*

File: app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+     * Unit test of fetchSubscriberDetails makes the call to profileInteractor execute method, will
+     * have response the profile information with has this response calls the repositoryInteractor
+     * execute method that will have response a list of the user repositories, first should show the loading
+     * indicator with the repositoryInteractor response dismiss the loading indicator and show the
+     * profile and repository information
+     *
+     * @throws Exception
+     */
+    @Test
+    public void fetchSubscriberDetails_withAttachedViewAndProfileInteractorResponse_ShouldCallRepositoriesInteractor_withRepositoriesInteractorResponse() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java]: fetchSubscriberDetails_withAttachedViewAndSuccessfulProfileAndRepositoryCalls_ShouldRenderProfileAndRepositories

File: app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        Profile mockProfile = new Profile("mockLogin", "www.mockulr.com", "www.mockurl.com", "mockName", "mockCompany", "mockLocation", 1, 1, 1);
+        List<Repository> mockUserRepositoryList = Arrays.asList(new Repository("mockName", "www.mockUrl.com"));
+        when(profileInteractor.execute(anyString())).thenReturn(Observable.just(mockProfile));
+        when(repositoriesInteractor.execute(anyString())).thenReturn(Observable.just(mockUserRepositoryList));
+        presenter.fetchSubscriberDetails(anyString());
+        verify(mockView, times(2)).render(any(SubscriberDetailsState.class));
+        verify(repositoriesInteractor, times(1)).execute(anyString());
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java]: By the default it should be 1

File: app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+     * @throws Exception
+     */
+    @Test
+    public void fetchSubscriberDetails_withAttachedViewAndProfileInteractorResponse_ShouldCallRepositoriesInteractor_withRepositoriesInteractorErrorResponse() throws Exception {
+        Profile mockProfile = new Profile("mockLogin", "www.mockulr.com", "www.mockurl.com", "mockName", "mockCompany", "mockLocation", 1, 1, 1);
+        when(profileInteractor.execute(anyString())).thenReturn(Observable.just(mockProfile));
+        when(repositoriesInteractor.execute(anyString())).thenReturn(Observable.<List<Repository>>error(new UnknownHostException()));
+        presenter.fetchSubscriberDetails(anyString());
+        verify(mockView, times(2)).render(any(SubscriberDetailsState.class));
+        verify(repositoriesInteractor, times(1)).execute(anyString());
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java]: By default it should be 1

File: app/src/test/java/com/globant/practice/presentation/presenter/SubscriberListPresenterTest.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    /**
+     * Unit test for the fetchUser method with a valid response of the web services, first should show
+     * the loading indicator when receive the mock list response should dismiss the loading indicator
+     * and show the subscribers
+     *
+     * @throws Exception
+     */
+    @Test
+    public void fetchUsers_withViewAttachedAndWebserviceResponse_ShouldShowLoadingIndicatorAndSubscribers() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/SubscriberListPresenterTest.java]: *fetchUsers_withViewAttachedAndSuccessfulResponse_ShouldRenderLoadingIndicatorAndSubscribers*

File: app/src/test/java/com/globant/practice/presentation/presenter/SubscriberListPresenterTest.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+    /**
+     * Unit test for the fetchUser method with a error response of the web services, first should show
+     * the loading indicator when receive the error response should dismiss the loading indicator
+     * and show the error message
+     *
+     * @throws Exception
+     */
+    @Test
+    public void fetchUsers_withViewAttachedAndWebserviceErrorResponse_ShouldShowLoadingIndicatorAndErrorMessage() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/SubscriberListPresenterTest.java]: *fetchUsers_withViewAttachedAndResponseFails_ShouldRenderLoadingIndicatorAndError*

File: app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * Unit test of loadWebPage, should show the loading indicator and initializes the load of
+     * the web page because the view is attached
+     *
+     * @throws Exception
+     */
+    @Test
+    public void loadWebPage_withAttachView_ShouldShowLoadingIndicatorAndLoadWebPage() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java]: withViewAttached

File: app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * Unit test of webPageLoadComplete, should dismiss the loading indicator because the view is
+     * attached
+     *
+     * @throws Exception
+     */
+    @Test
+    public void webPageLoadComplete_withAttachView_ShouldDismissLoadingIndicator() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java]: withViewAttached

File: app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+     * the web page because the view is attached
+     *
+     * @throws Exception
+     */
+    @Test
+    public void loadWebPage_withAttachView_ShouldShowLoadingIndicatorAndLoadWebPage() throws Exception {
+        presenter.attachView(mockView);
+        when(mockWebClientState.isLoading()).thenReturn(false);
+        presenter.loadWebPage("www.mockUrl.com");
+        verify(mockView, times(1)).render(any(WebClientState.class));
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java]: by default it should be 1

File: app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * Unit test of webPageError, should not dismiss the loading indicator and didn't show the error
+     * message because the view is detached
+     *
+     * @throws Exception
+     */
+    @Test
+    public void webPageError_witDetachView_ShouldNotDismissLoadingIndicator() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java]: You miss the **h** in with. Also, rename it to withViewDetached

File: app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * Unit test of webPageLoadComplete, should not dismiss the loading indicator because the view
+     * is detached
+     *
+     * @throws Exception
+     */
+    @Test
+    public void webPageLoadComplete_withDetachView_ShouldNotDismissLoadingIndicator() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java]: withViewDetached

File: app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * Unit test of the loadWebPage method, should not show the loading indicator and didn't load
+     * the web page because the view is detached
+     *
+     * @throws Exception
+     */
+    @Test
+    public void loadWebPage_withDetachView_ShouldNotShowLoadingIndicator() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/WebClientPresenterTest.java]: withViewDetached

**Review by daniellemariani — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java [hunk 14]
```diff
@@ -154,13 +154,13 @@ public void fetchSubscriberDetails_withAttachedViewAndProfileInteractorResponse_
      * @throws Exception
      */
     @Test
-    public void fetchSubscriberDetails_withAttachedViewAndProfileInteractorResponse_ShouldCallRepositoriesInteractor_withRepositoriesInteractorResponse() throws Exception {
+    public void fetchSubscriberDetails_withAttachedViewAndSuccessfulProfileAndRepositoryCalls_ShouldRenderProfileAndRepositories() throws Exception {
```
Comment (daniellemariani) [inline app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java]: withViewAttached

**Review by daniellemariani — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java (hunk 14 above)
Comment (jvc931) [reply app/src/test/java/com/globant/practice/presentation/presenter/SubscriberDetailsPresenterTest.java]: Done

**Review by daniellemariani — APPROVED** [review]
LG

**Pull request merged**